src/callbacks_step/amr_dg2d.jl

```diff
@@ -136,6 +140,106 @@ function refine!(u_ode::AbstractVector, adaptor, mesh::Union{TreeMesh{2}, P4estM
     return nothing
 end
 
+# Refine elements in the DG solver based on a list of cell_ids that should be refined.
+# If an element refines the solution scaled by the Jacobian `J*u` is interpolated from
+# the parent element into the four children elements. The solution on each child
+# element is then recovered by dividing by the new element Jacobians.
+function refine!(u_ode::AbstractVector, adaptor, mesh::P4estMesh{2},
+                 equations, dg::DGSEM, cache, elements_to_refine)
```

Comment:
It looks like most code is copy-pasted. Since this stuff is already ugly as hell, could we maybe try to reduce the amount of code we have to maintain? Two options are
- Add the new steps using `if` blocks to the old method
- Do some kind of scaling in both cases - using either helper functions that get specialized on the `mesh` type of `if` blocks

Reply:
Adding helper functions might be the way to go. In an old iteration I also had `TreeMesh` use the `J*u` interpolation / projection such that there was still only a single `coarsen` or `refine` function. This worked fine but would mean we need to update several more test values for the `TreeMesh` testing.

Reply:
I took a swing at adding `if` blocks checking whether or not we are on a `P4estMesh`. This allowed me to remove the duplicate functions and reunify everything into a single version of the `refine!` and `coarsen!` routines.